scripts/compaction_bounds.zig

```diff
@@ -0,0 +1,160 @@
+const std = @import("std");
+const assert = std.debug.assert;
+
+const Transfer = @import("../src/tigerbeetle.zig").Transfer;
+const IndexCompositeKeyType = @import("../src/lsm/groove.zig").IndexCompositeKeyType;
+const StateMachine = @import("../src/state_machine.zig").StateMachineType(
+    @import("../src/storage.zig").Storage,
+    .{
+        .message_body_size_max = config.message_body_size_max,
+    },
+);
+const config = @import("../src/config.zig");
+
+pub fn main() !void {
+    std.debug.print("Workload: 1_000_000 transfers/s.\n\n", .{});
+
+    const transfers_per_second = 1_000_000;
+    const transfers_per_batch = StateMachine.constants.batch_max.create_transfers;
+
+    const cases = .{
+        // The current code.
+        .current,
+        // Change level 0 to only contain 1 level.
+        .small_level_0,
+        // Change the immutable table to config.lsm_table_size_max.
+        // (Complicates compaction pacing.)
+        .large_immutable_table,
+        // Change the disk tables to match the size of the immutable table.
+        .small_disk_tables,
+        .small_level_0_and_small_disk_tables,
+    };
+    inline for (cases) |case| {
+        std.debug.print("{}:\n", .{case});
+
+        const existing_transfer_counts = [_]usize{
+            1_000_000,
+            10_000_000,
+            100_000_000,
+            1_000_000_000,
+        };
+        for (existing_transfer_counts) |existing_transfer_count| {
+            var read_bytes_per_second: usize = 0;
+            var write_bytes_per_second: usize = 0;
+
+            // Every transfer must be written to the WAL.
+            write_bytes_per_second += transfers_per_second * @sizeOf(Transfer);
```

Comment:
Nit: Committing these transfers also requires reads to the new account-object, account-id, and balance trees to check the debit/credit accounts and balances.

Reply:
If I model the worst-case for these reads I get:

```
> zig run scripts/compaction_bounds.zig --main-pkg-path ./
Workload: 1_000_000 transfers/s.

.current:
Existing transfers =    1000000. Read ~= 1253.63 gb/s. Write ~= 155.40 gb/s. Throughput at 4gb/s ~=    2839 transfers/s.
Existing transfers =   10000000. Read ~= 2352.83 gb/s. Write ~= 155.97 gb/s. Throughput at 4gb/s ~=    1594 transfers/s.
Existing transfers =  100000000. Read ~= 3457.09 gb/s. Write ~= 161.59 gb/s. Throughput at 4gb/s ~=    1105 transfers/s.
Existing transfers = 1000000000. Read ~= 4561.34 gb/s. Write ~= 167.22 gb/s. Throughput at 4gb/s ~=     846 transfers/s.

.small_level_0:
Existing transfers =    1000000. Read ~= 1118.01 gb/s. Write ~= 19.78 gb/s. Throughput at 4gb/s ~=    3516 transfers/s.
Existing transfers =   10000000. Read ~= 3321.46 gb/s. Write ~= 25.97 gb/s. Throughput at 4gb/s ~=    1195 transfers/s.
Existing transfers =  100000000. Read ~= 4425.72 gb/s. Write ~= 31.59 gb/s. Throughput at 4gb/s ~=     897 transfers/s.
Existing transfers = 1000000000. Read ~= 5529.98 gb/s. Write ~= 37.22 gb/s. Throughput at 4gb/s ~=     718 transfers/s.

.large_immutable_table:
Existing transfers =    1000000. Read ~= 1103.63 gb/s. Write ~=  5.86 gb/s. Throughput at 4gb/s ~=    3605 transfers/s.
Existing transfers =   10000000. Read ~= 2202.83 gb/s. Write ~=  6.43 gb/s. Throughput at 4gb/s ~=    1811 transfers/s.
Existing transfers =  100000000. Read ~= 3307.09 gb/s. Write ~= 12.05 gb/s. Throughput at 4gb/s ~=    1205 transfers/s.
Existing transfers = 1000000000. Read ~= 4411.34 gb/s. Write ~= 17.68 gb/s. Throughput at 4gb/s ~=     903 transfers/s.

.small_disk_tables:
Existing transfers =    1000000. Read ~= 2200.10 gb/s. Write ~=  3.24 gb/s. Throughput at 4gb/s ~=    1815 transfers/s.
Existing transfers =   10000000. Read ~= 3300.23 gb/s. Write ~=  4.73 gb/s. Throughput at 4gb/s ~=    1210 transfers/s.
Existing transfers =  100000000. Read ~= 4400.36 gb/s. Write ~=  6.23 gb/s. Throughput at 4gb/s ~=     908 transfers/s.
Existing transfers = 1000000000. Read ~= 5500.49 gb/s. Write ~=  7.73 gb/s. Throughput at 4gb/s ~=     726 transfers/s.

.small_level_0_and_small_disk_tables:
Existing transfers =    1000000. Read ~= 3299.06 gb/s. Write ~=  3.57 gb/s. Throughput at 4gb/s ~=    1211 transfers/s.
Existing transfers =   10000000. Read ~= 4399.19 gb/s. Write ~=  5.07 gb/s. Throughput at 4gb/s ~=     908 transfers/s.
Existing transfers =  100000000. Read ~= 5499.32 gb/s. Write ~=  6.56 gb/s. Throughput at 4gb/s ~=     726 transfers/s.
Existing transfers = 1000000000. Read ~= 6599.46 gb/s. Write ~=  8.06 gb/s. Throughput at 4gb/s ~=     605 transfers/s.
```

The average case should be much better, but I don't know how to estimate it accurately.

Reply:
:thinking: I don't understand why this is so extreme; can I see the diff?

Reply:
Nevermind, just saw it posted.

Reply:
Probably many lookups will be cached, and probably we won't have to read every level on most uncached lookups. But N * uncached lookups is really expensive relative to N * transfer writes.

Reply:
Yeah, it also assumes that the bloom filters always returned false positives, and that everything was on the last level — definitely the worst case, but not likely.